### ISSUE TITLE
Match guard can both move and static-promote a single constant

### ISSUE THREAD

**Issue by theemathas** [rust-lang/rust#145237]
I'm not sure if this is a bug or not.

I expected the below code to produce a compile error, but instead it compiles fine.

```rust
struct Thing;

fn foo() -> &'static Thing {
    let y;
    match Thing {
        x if { y = &x; true } => {
            // This drop should invalidate y
            drop(x);
            y
        }
        _ => {
            panic!();
        }
    }
}
```

It seems that the single `Thing` constant is both moved and static-promoted. This seems very strange.

Related to #144939. Discovered in https://github.com/rust-lang/rust/pull/141295#issuecomment-3173106871.

@rustbot labels +A-patterns +A-const-eval

### Meta

Reproduces on the playground with version `1.91.0-nightly (2025-08-09 ca77504943887037504c)`

**Comment by kakserpom** [comment]
It doesn't compile if you add
```rust
impl Drop for Thing {
    fn drop(&mut self) {}
}
```

```
error[E0505]: cannot move out of value because it is borrowed
  --> src/main.rs:8:9
   |
8  |         x if {
   |         ^
   |         |
   |         move out of value occurs here
   |         borrow of value occurs here
...
15 |             y
   |             - returning this value requires that borrow lasts for `'static`
```

**Comment by kakserpom** [comment]
How about these? I got rid of 'static.

- (FAILS) [impl Drop AND drop()](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=bee0eefe229f7c0891ddf3ed2f4f6e75)
- (FAILS)[impl Drop](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=d22a571fd2e4f7adb70a0c2d7e310525)
- (OK) [drop()](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=976b781c68d6efe2e490cbe3b326fe8e)

**Comment by theemathas** [comment]
`drop()` here can't change run time behavior, but it can change whether code compiles or not. I included it here as a stand-in for any operation that consumes `x`, to make it clearer that the code should not compile.

**Comment by dianne** [comment]
This does feel strange. Given that `x` is bound by-move, I'd expect there to be a move error from the scrutinee being moved into `x` after the guard succeeds. I feel like this may be a T-lang question: either this behavior should be specified as part of the language, or the language as defined by the compiler is too permissive.

@rustbot label +T-lang +I-lang-nominated

**Comment by Nadrieril** [comment]
I'm surprised this works even without the drop. We don't allow `let x = Thing; let y = &x;` to static-promote, do we?

**Comment by traviscross** [comment]
**Analysis**: First, let's consider the relevant Reference text:

> r[expr.match.scrutinee-value]
> If the scrutinee expression is a [value expression], it is first evaluated into a temporary location, and the resulting value is sequentially compared to the patterns in the arms until a match is found...
>
> r[expr.match.guard.shared-ref]
> Before evaluating the guard, a shared reference is taken to the part of the scrutinee the variable matches on. While evaluating the guard, this shared reference is then used when accessing the variable.
>
> r[expr.match.guard.value]
> Only when the guard evaluates to true is the value moved, or copied, from the scrutinee into the variable. This allows shared borrows to be used inside guards without moving out of the scrutinee in case guard fails to match.
>
> r[destructors.scope.const-promotion]
> Promotion of a value expression to a `'static` slot occurs when the expression could be written in a constant and borrowed, and that borrow could be dereferenced where the expression was originally written, without changing the runtime behavior. That is, the promoted expression can be evaluated at compile-time and the resulting value does not contain [interior mutability] or [destructors] (these properties are determined based on the value where possible, e.g. `&None` always has the type `&'static Option<_>`, as it contains nothing disallowed).

Now, let's consider:

```rust
fn _f1() {
    match None::<String> {
        x if { let _: &'static _ = &x; true } => {}
        _ => {}
    }
}
```

Here's what I think is happening:

First, the `None::<String>` value expression is evaluated and stored as a temporary.

Then, due to the presence of a guard, "a shared reference is taken to the part of the scrutinee the variable matches on".  Here, that's the entire temporary.  That borrow triggers const promotion -- not the borrow in the guard.  We can see this if we look at the MIR for:

```rust
fn _f2() {
    match None::<String> {
        x if true => {}
        _ => {}
    }
}
```

Const promotion happens there too, simply due to the presence of the trivial guard.

Is this const promotion valid?  To determine that, we ask:

- Can the value expression be evaluated at compile-time? Yes.
- Does the resulting value contain interior mutability? No.
- Does the resulting value have a non-trivial destructor? No.

Therefore, "the expression could be written in a constant and borrowed, and that borrow could be dereferenced where the expression was originally written, without changing the runtime behavior."  The const promotion seems valid.

But consider:

```rust
fn _f3() {
    let slot = None::<String>;
    if {
        let slot_ref = &slot;
        //~^ ERROR does not live long enough
        let _: &'static _ = &*slot_ref;
        true
    } {
        let x = slot;
        //~^ ERROR cannot move out of `slot` because it is borrowed
    }
}
```

How do we explain that?  That should be roughly equivalent to `_f1`, right?  Why don't we do promotion there?

For this, we need to look at the [const-eval](https://github.com/rust-lang/const-eval) document chapter on [promotion](https://github.com/rust-lang/const-eval/blob/master/promotion.md).  As it describes:

> Promotable expressions cannot refer to named locals. This is not a technical limitation with the CTFE engine. While writing `let x = {expr}` outside of a const context, the user likely expects that `x` will live on the stack and be initialized at run-time.  Although this is not (to my knowledge) guaranteed by the language, we do not wish to violate the user's expectations here.

This is where our attempt to simulate what `match` is doing falls short.  We've now named the local, and this rule now prevents promotion.  The temporary for the scrutinee, on the other hand, isn't named, and so we allow promotion.

Net-net, I suspect the behavior is valid and is probably desirable.  It goes all the way back to Rust 1.39 and <https://github.com/rust-lang/rust/pull/63118>.

**Comment by theemathas** [comment]
> First, the None::<String> value expression is evaluated and stored as a temporary.

I have some doubts about the "stored as a temporary" step actually happening precisely here. The following code compiles and prints nothing when ran.

```rust
struct LoudDrop;
impl Drop for LoudDrop {
    fn drop(&mut self) {
        println!("drop");
    }
}

static A: LoudDrop = LoudDrop;

fn main() {
    'a: {
        match A {
            x if {
                let y: &'static _ = &x;
                assert!(std::ptr::eq(y, &A));
                break 'a;
            } => {}
            _ => {}
        }
    }
}
```

**Comment by traviscross** [comment]
The `A` there is not a value expression, so the appropriate Reference language for that is:

> r[expr.match.scrutinee-place]
> When the scrutinee expression is a [place expression], the match does not allocate a temporary location;
however, a by-value binding may copy or move from the memory location. When possible, it is preferable to match on place expressions, as the lifetime of these matches inherits the lifetime of the place expression rather than being restricted to the inside of the match.

**Comment by dianne** [comment]
Something interesting about this being a consequence of [`expr.match.guard.shared-ref`](https://doc.rust-lang.org/reference/expressions/match-expr.html#r-expr.match.guard.shared-ref): that's a place where the Reference and compiler don't totally agree. The Reference describes how it works for by-value bindings, but for by-ref bindings, the compiler currently eagerly creates the binding used in the arm, then takes a shared reference to that to be used in the guard. For by-shared-ref bindings, the end result in this case is about the same:
```rust
fn main() {
    // This works: the `ref x` binding results in promotion,
    // regardless of whether or not there's a guard
    match None::<String> {
        ref x if { let _: &'static _ = x; true } => {}
        _ => {}
    }
}
```
but for by-mutable-ref bindings, it's easier to observe the difference:
```rust
fn main() {
    // This doesn't work: the `ref mut x` binding doesn't promote,
    // regardless of whether or not there's a guard
    match None::<String> {
        ref mut x if { let _: &'static _ = x; true } => {}
        _ => {}
    }
}
```

**Comment by RalfJung** [comment]
That's by design. We do not want to promote this:
```rust
let x = 0;
let y = &x
```

**Comment by traviscross** [comment]
>> An attempt at a statement: we shouldn't be const-promoting anything that you couldn't wrap in a const { … } block.
> 
> I agree with that. Every case where we violate that is a bug, IMO.

Would we expect this,

```rust
fn _f4() {
    match const { None::<String> } {
        x if { let _: &'static _ = &x; true } => (),
        _ => unreachable!(),
    }
}
```

to work?  On the one hand, we don't accept this,

```rust
    let x = const { None::<String> };
    let _: &'static _ = &x; //~ ERROR
```

but on the other, that's not what the `match` is doing.  The `&x` in the guard is a reference to the constant, not to the place of `x` -- we haven't moved into `x` yet.

For my part, I probably think we should accept `_f4`, and if we do, then that provides a good story for not accepting the version without the explicit `const { .. }`, as people can then just add that.

**Comment by RalfJung** [comment]
> What are your thoughts on that?

Maybe we should reimplement promotion on HIR/THIR but, well, currently it acts on MIR. @oli-obk knows more about the history here; IIRC it used to be on HIR but got moved to MIR for some reason.

> The promotion seemingly happens due to 1) the scrutinee being a value expression which 2) causes that expression to be evaluated to a temporary, then 3) the presence of a guard (any guard at all) triggers a shared borrow of that temporary to be taken, and 4) as (all of:) we're not borrowing a named local, we can evaluate it at compile-time, and it does not have a non-trivial destructor or interior mutability, we promote it.

Yeah, that sounds plausible. We should distinguish the guard borrows from normal borrows of temporaries to prevent this.

> Would we expect this

I don't think so. I think `const { None::<String> }` and `None::<String>` should be equivalent -- both are just consts.

We should only promote if there's literally an `&<promoteable expr>` (or the equivalent with `ref` I guess, if we can't take that one back).

> but on the other, that's not what the match is doing. The &x in the guard is a reference to the constant, not to the place of x -- we haven't moved into x yet.

It's a reference to the place we re matching on, which place the constant has been written into. But it's indirected by a binder. I am not convinced we should promote through such indirections, that seems awfully non-local.

**Comment by traviscross** [comment]
> I don't think so. I think `const { None::<String> }` and `None::<String>` should be equivalent -- both are just consts.

Notably, in the MIR, we only produce an actual constant if `const { .. }` is used.

Would you expect this

```rust
fn _f5() {
    const X: Option<String> = None::<String>;
    let _: &'static _ = match X { ref x => x };
}
```

to fail as well?  That working is what leads to my intuition for why this

```rust
fn _f6() {
    let _: &'static _ = match const { None::<String> } { ref x => x };
}
```

should work, and that backs my intuition for why this

```rust
fn _f4a() {
    const X: Option<String> = None::<String>;
    match X {
        x if { let _: &'static _ = &x; true } => (),
        _ => unreachable!(),
    }
}
```

and

```rust
fn _f4b() {
    match const { None::<String> } {
        x if { let _: &'static _ = &x; true } => (),
        _ => unreachable!(),
    }
}
```

should work.

**Comment by steffahn** [comment]
@theemathas 

> A somewhat related weirdness *[…]*

On one hand this seems related of course. (Let me rewrite it slightly, anyway:)
```rs
fn main() {
    let y;

    match 1 {
        mut x if {y = &x; true} => {
            println!("{x}");
            x = 2;
            println!("{x}");
            println!("{y}");
        }
        _ => {}
    }
}
```
[↝ playground ⮺](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=2ee2a5a2b5513d25d8b41da6985fd55d)

on the other hand the thing that went wrong here - a combination of mutability, ergonomic features[^1] meant for movable types and then usage of a `Copy` one after all, not playing well together; seems like a re-occurring issue to me as well! Specifically thus, I feel in some ways that code is also quite similar to this code:
```rs
fn main() {
    let mut x = 1;
    let y = &x;

    (move || {
        println!("{x}");
        x = 2;
        println!("{x}");
        println!("{y}");
    })();
}
```
[↝ playground ⮺](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=11521ccfd77fe9d6cc19e8e1e0c5babc)

[^1]: In this case, the ergonomics are from how guards maintain the ability to fail late – after matching against `mut x` already – without leaving the expression being matched on uninitialized (if it wasn’t a `Copy` type, anyway) – by means of faking things by something like
* match against "a `ref r_x` pattern first", then "run the guard in a mode ~~as cursed as C++ references~~ that is ~~as cursed as closures~~ implicitly turning mentions of `x` into `*r_x`" and finally "actually do the matching" (for non-`Copy` types this would end the lifetime of the `&x` inside the guard; you could still observe that it [changed locations](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=8f68a225c687d97818df23760430d57a), anyway).
